### PR TITLE
Fix deduction deserializer with DefaultTypeResolverBuilder (pending #)

### PR DESCRIPTION
**Pull request by Bluexin**
Fix for https://github.com/FasterXML/jackson-module-kotlin/issues/568, which will be transferred to this repo (if you prefer I can just create a new issue instead) and then I will update the comments in the test.
I will also send you the signed CLA shortly.

### CODE REVIEW

**Comment by cowtowncoder** [comment]
Excellent, thank you @Bluexin! The only other thing I need now is CLA (unless you have sent one earlier, if so let me know). It's here:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

and the usual way is to print it, fill & sign, scan/take pic, email to `info` at fasterxml dot com.
This is one-time thing so CLA is good for any and all future contributions.

Once I get that I can merge this in for inclusion in 2.14 assuming tests pass.

**Comment by cowtowncoder** [comment]
@Bluexin um, did you do the usual `mvn clean test` verification with this patch? Looks like this breaks a whole lot of test cases. I think the logic defined was there for a reason and looking at it, there are two cases:

1. Explicit polymorphic handling (via `@JsonTypeInfo`)
2. "Default typing" with no annotations, applied to type categories. Uses (always?) class name as Type Id

In case (2) subtype information should not be needed; I suspect that might break handling.
I do not remember all details, nor did check failure cases yet.

**Comment by Bluexin** [comment]
I was having trouble with some of the test cases even before changing any code, and running a full maven test (or "all tests" in IDEA) showed many failing/erroring but then running the test classes one by one they all passed except for one expecting a class to be missing that wasn't actually missing. So I figured something might be up with the JDK version I was using or something else with my setup.
I will have another look, but there was definitely something weird so it might take some time.

I sent the CLA last night.

**Comment by cowtowncoder** [comment]
@Bluexin I do not usually run all test from IDE; part of the reason is that `mvn test` will skip tests under `..../failing/` -- but IDE has no idea that these should not be run. They are runnable from IDE (not annotated to be excluded) to allow manual usage.
This might explain what you see.

But I don't think this fix can be used as is, after all.

**Comment by cowtowncoder** [comment]
CI still reports tons of failures:

```
Tests run: 3694, Failures: 51, Errors: 4, Skipped: 0
```

so I can't really proceed with this until those are better understood.

**Comment by Bluexin** [comment]
Sorry I just got back to this.
After some more digging into this, it turns out the addition of the test is what is causing all these failures somehow : 
I tried without the changes but with the tests and saw all these failures, as well as with the changes but without the test and saw no failures.
I'm tryin gto find out why this happens now.

**Comment by Bluexin** [comment]
Turns out the test I copied as structure was using a shared test mapper, which meant registering custom types for it would skew other tests results.
Got all green locally, I hope the CI build will be happy as well :)

File: src/test/java/com/fasterxml/jackson/databind/deser/TestDeserializerFactoryWithDefaultTypeResolverBuilder.java
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/deser/TestDeserializerFactoryWithDefaultTypeResolverBuilder.java:65]: Should this verify the result just in case? (I realize that for specific issue a NPE was thrown but code could theoretically, say, return `null`)

**Comment by cowtowncoder** [comment]
@Bluexin ah. That makes sense. Thank you for following up; I should have caught that. It is kind of obscure issue.

I wish this came up 1 week ago before 2.14.0-rc1 so we could get some more testing but... oh well. I am already thinking of releasing 2.14.0-rc2.

So I think I'll merge this later today or tomorrow, after thinking it through once more.

File: src/test/java/com/fasterxml/jackson/databind/deser/TestDeserializerFactoryWithDefaultTypeResolverBuilder.java
Comment (Bluexin) [reply src/test/java/com/fasterxml/jackson/databind/deser/TestDeserializerFactoryWithDefaultTypeResolverBuilder.java]: You're absolutely right, I added some extra assertions

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
@Bluexin no problem, we all have other important things to do. I appreciate you getting this done whenever you could!

**Comment by Bluexin** [comment]
Oh I just realised the issue never got moved from https://github.com/FasterXML/jackson-module-kotlin/issues/568 so the test doesn't mention any issue number, I hope that's alright

**Comment by cowtowncoder** [comment]
No problem @Bluexin I can just use PR id instead of issue; have done it in the past.

**Pull request merged**
